src/append_only_data.rs

```diff
@@ -658,15 +658,39 @@ impl UnseqAppendOnlyData<UnpubPermissions> {
     }
 }
 
+fn check_dup(data: &[(Vec<u8>, Vec<u8>)], entries: &mut Entries) -> Result<()> {
+    let new: BTreeSet<&Vec<u8>> = entries.iter().map(|(key, _value)| key).collect();
+
+    // If duplicate entries are present in the push.
+    if new.len() < entries.len() {
+        return Err(Error::DuplicateEntryKeys);
+    }
+
+    let existing: BTreeSet<&Vec<u8>> = data.iter().map(|(key, _value)| key).collect();
+    if !existing.is_disjoint(&new) {
+        let dup: Entries = entries
+            .drain(..)
+            .filter(|(key, _value)| existing.contains(&key))
+            .collect();
+        return Err(Error::KeysExists(dup));
+    }
+    Ok(())
+}
+
 impl<P> SeqAppendOnly for SeqAppendOnlyData<P>
 where
     P: Permissions + std::hash::Hash + Clone,
 {
-    fn append(&mut self, entries: &[(Vec<u8>, Vec<u8>)], last_entries_index: u64) -> Result<()> {
+    fn append(&mut self, mut entries: Entries, last_entries_index: u64) -> Result<()> {
+        match check_dup(&self.inner.data, entries.as_mut()) {
```

Comment:
Probably can be replaced with just
```suggestion
        if let Err(e) = check_dup(&self.inner.data, entries.as_mut()) {
            return Err(e);
        }
```

Reply:
Or even just
```rust
check_dup(&self.inner.data, entries.as_mut())?;
```